rfc/languages-in-w3s.md

```diff
@@ -0,0 +1,84 @@
+- By and large, people get stuff done quickly in Go because Go is extremely simple. It's very fast to learn, easy to read, and it all looks the same. The only problem is its extremely boring. (Note from Hannah: When I first started writing Go I hated it. I thought someone took the joy out of programming. Now I love it cause I get my joy from what I'm building, not my programming language)
+- Lots and lots of Filecoin tooling and software is written in Go. In fact, W3S already has Go in it, we just don't maintain it -- it's called Spade
+- Garbage collection and green threads present challenges for using Go over an FFI, or for compiling go to WebAssembly, though TinyGo has improved this story by reducing the runtime size.
+
+### Rust
+
+Caveat from Hannah: I am not a highly experienced Rust dev
+
+- Rust is a better C/C++. Like definitely much better. This is huge cause no one wrote a better C/C++ for low level programming for 20 years.
+- However, by and large, you should probably not write programs in Rust unless you'd consider writing them in C/C++. There is minimal evidence one can be as productive in Rust as higher level languages.
+- In particular, you should probably not write web servers in Rust or any other asynchronous programs unless you have extra time on your hands which we don't.
+- BTW, if you haven't programmed in C/C++ recently, you've probably forgotten that long compile times are a thing on large projects in low level languages. 
+- Rust is essential for highly performant synchronous compute tasks. That's why people use it for smart contracts, hashing, and proofs.
+- Rust has the best WebAssembly story of any language.
+
+## Table Stakes For Considering A Language for W3S
+
+In addition to having a compelling engineering or business reason, we have some other things to consider before we adopt a language:
+
+- UCAN libraries and tooling especially for anything with a web interface
+- IPLD/IPFS/Libp2p libraries (we may not always need libp2p going forward)
+
+## Proposing a language to adopt
+
+- The best time to propose a language to adopt is at the start of a new project
+- Present a proposal to write a project in a new language via an RFC
+   - Try to objectively present business and engineering tradeoffs of writing the project in a language we already support vs writing it in a new language
+   - Identify gaps in language support or tooling related to our software stack and present a plan to address it
+   - Ideally, make a case for why this will be useful for other projects going forward
+   - Present a plan for bringing other team members up to speed on the language you propose to use. Usually the language proposer will have to captain the process of language education.
+- Since it's a big decision, we should try to get input from all devs before making a decision at least till our team gets bigger
+- Must have sign off from senior engineering leadership (for now Hannah and Alan)
```

Comment:
my concern with Rust has a lot to do with the speed of development for new code, as opposed to optimizing already well prototyped services.

Reply:
maybe @mroth wants to weight in here?

Reply:
One concern with Rust for larger projects is that it can slow down iterative development. There is more code, due to the need to precisely manage resources, among other things, and refactoring tends to be more difficult. With Go, it is relatively easy to refactor as a project evolves, allowing for rapid iteration of development cycles.

In addition to longer refactor time, rust compile times are significantly longer than with Go. The difference becomes pronounced as the amount of code and dependencies grows. This makes testing and trying out new code quicker in Go, resulting in considerable time savings over the course of development.

This echos what was already said by @hannahhoward, and is in the spirit of "consider rust when you'd consider C/C++".

Rust is an awesome language and enables one to write highly optimal and safe code. However, I think it is a mistake to fall in love with this without also understanding the value of what Go beings. That is: Go does everything it can to minimize thought and time spent getting the compiler to do exactly what you want, allowing you to focus on best expressing the logic needed to solve whatever problem you are trying to solve. There is some cost to that (like GC), but the benefits of productivity and clear maintainable code very often far outweigh that cost.

Reply:
> refactoring tends to be more difficult

It depends. Sometimes a structural change is difficult due to the ownership model. On the other hand, sometimes a refactor is easy because, once something compiles, it usually works.

Bug I agree that iteration cycles in rust tend to be slower. E.g., I wouldn't prototype a project with an _unknown_ architecture in rust if I could help it. I'd start by prototyping and answering all the architectural questions in Go, then I'd rewrite it in rust.

---

One more thing for rust, however: C compat. CGO is a nightmare.